inhibit.go

```diff
@@ -125,20 +177,48 @@ func NewInhibitRule(cr *config.InhibitRule) *InhibitRule {
 		SourceMatchers: sourcem,
 		TargetMatchers: targetm,
 		Equal:          equal,
+		scache:         map[model.Fingerprint]*types.Alert{},
 	}
 }
 
-// Mutes returns true iff the Inhibition rule applies for the given
-// source and target label set.
-func (r *InhibitRule) Mutes(source, target model.LabelSet) bool {
-	if !r.TargetMatchers.Match(target) || !r.SourceMatchers.Match(source) {
-		return false
-	}
-	for ln := range r.Equal {
-		if source[ln] != target[ln] {
-			return false
+// set the alert in the source cache.
+func (r *InhibitRule) set(a *types.Alert) {
+	r.mtx.Lock()
+	r.mtx.Unlock()
+
+	r.scache[a.Fingerprint()] = a
+}
+
+// hasEqual checks whether the source cache contains alerts matching
+// the equal labels for the given label set.
+func (r *InhibitRule) hasEqual(lset model.LabelSet) bool {
+	r.mtx.RLock()
+	defer r.mtx.RUnlock()
+
+Outer:
+	for _, a := range r.scache {
+		// The cache might be stale and contain resolved alerts.
+		if a.Resolved() {
+			continue
 		}
+		for n := range r.Equal {
+			if a.Labels[n] != lset[n] {
+				continue Outer
```

Comment:
But it would be wrong. We have to check the subsequent alerts as well as one is sufficient to mute the target alert.